### PR TITLE
[mqtt] Add ClientInfo to SessionState so that Offline sessions can have a last known ClientInfo

### PR DESCRIPTION
**Pull request by dylanbronson**
The significant change here is adding ClientInfo to SessionState, so that Offline sessions will have the "last known" client_info.
Now broker can reevaluate offline sessions when an auth update comes in

Everything else is plumbing.

### CODE REVIEW

File: mqtt/mqtt-broker/src/proptest.rs [hunk 1]
```diff
@@ -53,6 +54,7 @@ prop_compose! {
     ) -> SessionSnapshot {
         SessionSnapshot::from_parts(
             client_id,
+            ClientInfo::new(SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 8080), AuthId::from("authId")),
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/proptest.rs:57]: replace with `arb_client_info()` ?

File: mqtt/mqtt-broker/src/state_change.rs [hunk 2]
```diff
@@ -100,15 +103,19 @@ mod tests {
...
         subscription::{Subscription, TopicFilter},
         tests::peer_addr,
-        Auth, AuthId, BrokerConfig, ClientId, ConnReq, SessionConfig,
+        Auth, AuthId, BrokerConfig, ClientId, ClientInfo, ConnReq, SessionConfig,
     };
 
     #[test]
     fn test_subscriptions() {
         let expected_id: ClientId = "Session".into();
+        let client_info = &ClientInfo::new(
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/state_change.rs:112]: IMO it's a bit confusing to create a ref. Create a value and pass it as a ref where needed.
```suggestion
        let client_info = ClientInfo::new(
```

**Review by vadim-kovalyov — COMMENTED** [review]
:+1:

File: mqtt/mqtt-broker/src/proptest.rs (hunk 1 above)
Comment (dylanbronson) [reply mqtt/mqtt-broker/src/proptest.rs:57]: oh yeah haha. Originally had as a placeholder. I'll fix

File: mqtt/mqtt-broker/src/broker.rs [hunk 3]
```diff
@@ -109,17 +109,16 @@ where
...
-                    qos: *sub.max_qos(),
-                });
-                activities.push((
-                    client_id.clone(),
-                    Activity::new(client_id.clone(), client_info.clone(), operation),
-                ));
-            }
+            let client_info = session.client_info();
+            let sub_topic_filter = sub.filter().to_string();
+            let operation = Operation::new_subscribe(proto::SubscribeTo {
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:114]: this is a bit weird: to create `proto::SubscribeTo` only to later destroy it and create internal representation :(

File: mqtt/mqtt-broker/src/broker.rs [hunk 4]
```diff
@@ -109,17 +109,16 @@ where
             .into_iter()
             .flat_map(HashMap::values)
         {
-            if let Ok(client_info) = session.client_info() {
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:112]: nit:
```rust
fn prepare_activities(client_id: &ClientId, session: &Session) -> Vec<(ClientId, Activity)> {
        session
            .subscriptions()
            .into_iter()
            .flat_map(HashMap::values)
            .map(|sub| {
                let operation = Operation::new_subscribe(proto::SubscribeTo {
                    topic_filter: sub.filter().to_string(),
                    qos: *sub.max_qos(),
                });

                let client_info = session.client_info().clone();
                let activity = Activity::new(client_id.clone(), client_info, operation);

                (client_id.clone(), activity)
            })
            .collect()
    }
```

File: mqtt/mqtt-broker/src/broker.rs [hunk 5]
```diff
@@ -701,6 +700,7 @@ where
 
     fn open_session(&mut self, auth_id: AuthId, connreq: ConnReq) -> Result<OpenSession, Error> {
         let client_id = connreq.client_id().clone();
+        let client_info = ClientInfo::new(connreq.peer_addr(), auth_id.clone());
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:703]: lets move this line closer to usage?

File: mqtt/mqtt-broker/src/lib.rs [hunk 6]
```diff
@@ -81,7 +81,7 @@ impl Display for ClientId {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 pub struct ClientInfo {
     peer_addr: SocketAddr,
     auth_id: AuthId,
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/lib.rs:87]: since session will always have last used client info, lets add client_id into this struct? it should simplify code a little bit

File: mqtt/mqtt-broker/src/session/offline.rs [hunk 7]
```diff
@@ -23,6 +23,10 @@ impl OfflineSession {
         self.state.client_id()
     }
 
+    pub fn client_info(&self) -> &ClientInfo {
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/session/offline.rs:26]: q: is it only for me, or it is a bit confusing to have `client_info` for Offline session? worth to rename to `last_client_info` or something like that?

**Review by dmolokanov — CHANGES_REQUESTED** [review]
Last time when I changed this code it was a hiccup to not re-use auth_id for newly created session if there was session before (active or non-active). Please double check all branches.

File: mqtt/mqtt-broker/src/session/offline.rs (hunk 7 above)
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session/offline.rs:26]: don't have strong preferences. We can rename to `last_client_info`

File: mqtt/mqtt-broker/src/broker.rs (hunk 5 above)
Comment (dylanbronson) [reply mqtt/mqtt-broker/src/broker.rs:703]: We use client_info within the match block multiple times. We could recreate it a few times if that's better, but I prefer like this

File: mqtt/mqtt-broker/src/broker.rs (hunk 3 above)
Comment (dylanbronson) [reply mqtt/mqtt-broker/src/broker.rs]: When do we destroy it later?

File: mqtt/mqtt-broker/src/persist.rs [hunk 8]
```diff
@@ -544,11 +544,13 @@ mod tests {
...
     use crate::{
-        persist::{ConsolidatedState, FileFormat, FilePersistor, Persist, VersionedFileFormat},
+        persist::{FileFormat, FilePersistor, Persist, VersionedFileFormat},
         proptest::arb_broker_snapshot,
         BrokerSnapshot,
     };
 
+    use super::ConsolidatedState;
+
     proptest! {
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/persist.rs:554]: nit: not sure why this change?

File: mqtt/mqtt-broker/src/proptest.rs [hunk 9]
```diff
@@ -112,6 +113,22 @@ prop_compose! {
...
+    pub fn arb_client_info()(
+        client_id in arb_clientid(),
+        auth_id in arb_auth_id(),
+        ip in arb_ip(),
+        port in arb_port(),
+    ) -> ClientInfo {
+        // Unfortunately we can't just call SocketAddr::arbitrary() because when serde occurs on SocketAddr,
+        // we lose the flowid and scope_id. They get set to 0 by default.
+        // This workaround manually sets them to 0 but uses arbitrary values for ip and port
+        // Issue opened: https://github.com/servo/bincode/issues/354
```
Comment (vadim-kovalyov) [inline mqtt/mqtt-broker/src/proptest.rs:126]: :+1: thanks for posting an issue

**Review by vadim-kovalyov — APPROVED** [review]
LGTM! :rocket:

**Review by vadim-kovalyov — APPROVED** [review]

File: mqtt/mqtt-broker/src/broker.rs [hunk 10]
```diff
@@ -807,23 +808,24 @@ where
...
-                    );
-                    let new_session = Session::new_persistent(auth_id, connreq, state);
-                    (new_session, true)
-                } else {
-                    info!("cleaning session for {}", client_id);
-                    let state = SessionState::new(client_id.clone(), self.config.session().clone());
-                    let new_session = Session::new_transient(auth_id, connreq, state);
-                    (new_session, false)
-                };
+            let (state, _will, handle) = current_connected.into_parts();
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:811]: On this line, we take the whole state with client_info and pass it for a connecting client which may have different addresses and credentials. I think we should create a new ClientInfo for the connecting client and then update state.

Since we didn't catch it with unit tests, we should add it

Comment (dmolokanov) [reply mqtt/mqtt-broker/src/broker.rs:811]: Scenarion:
1: client connects `{ client_id: "foo", "username": "foo", addr: 198.0.0.1 }`
state has session with client_info { client_id: "foo", "username": "**foo**", addr: **198.0.0.1** }
2: client connects `{ client_id: "foo", "username": "bar", addr: 198.0.0.2 }`
expected: 
state has session with client_info { client_id: "foor", "username": "**bar**", addr: **198.0.0.2** }
actual: 
state has session with client_info { client_id: "foor", "username": "**foo**", addr: **198.0.0.1** }

File: mqtt/mqtt-broker/src/broker.rs [hunk 11]
```diff
@@ -807,23 +808,24 @@ where
...
+            let old_session = Session::new_disconnecting(state.client_info().clone(), None, handle);
+            let (new_session, session_present) = if let proto::ClientId::IdWithExistingSession(_) =
+                connreq.connect().client_id
+            {
+                debug!(
+                    "moving persistent session to this connection for {}",
+                    client_id
+                );
+                let new_session = Session::new_persistent(auth_id, connreq, state);
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:820]: since we have auth_id in the state, should we simplify this a bit?
```suggestion
                let new_session = Session::new_persistent(connreq, state);
```

File: mqtt/mqtt-broker/src/broker.rs [hunk 12]
```diff
@@ -807,23 +808,24 @@ where
...
+                (new_session, true)
+            } else {
+                info!("cleaning session for {}", client_id);
+                let state =
+                    SessionState::new(state.client_info().clone(), self.config.session().clone());
+                let new_session = Session::new_transient(auth_id, connreq, state);
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:826]:
```suggestion
                let new_session = Session::new_transient(connreq, state);
```
same here

File: mqtt/mqtt-broker/src/proptest.rs [hunk 13]
```diff
@@ -140,6 +157,18 @@ pub fn arb_clientid() -> impl Strategy<Value = ClientId> {
     "[a-zA-Z0-9_()!@%,'=\\*\\$\\?\\-]{1,23}".prop_map(Into::into)
 }
 
+pub fn arb_auth_id() -> impl Strategy<Value = AuthId> {
+    "[a-zA-Z0-9]{1,23}".prop_map(AuthId::from)
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/proptest.rs:161]: what about an anonymous case?

**Review by vadim-kovalyov — APPROVED** [review]

File: mqtt/mqtt-broker/src/session/state/mod.rs [hunk 14]
```diff
@@ -108,6 +112,10 @@ impl SessionState {
         &mut self.waiting_to_be_sent
     }
 
+    pub fn set_client_info(&mut self, client_info: ClientInfo) {
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/session/state/mod.rs:115]: Could you move this setter closer and place it under fn client_info?

File: mqtt/mqtt-broker/src/broker.rs [hunk 15]
```diff
@@ -806,35 +806,36 @@ where
...
-            let old_session = Session::new_disconnecting(state.client_info().clone(), None, handle);
-            let (new_session, session_present) = if let proto::ClientId::IdWithExistingSession(_) =
-                connreq.connect().client_id
-            {
+        let (mut state, _will, handle) = current_connected.into_parts();
+        let old_session = Session::new_disconnecting(state.client_info().clone(), None, handle);
+        let client_id = connreq.client_id().clone();
+        let new_client_info = ClientInfo::new(client_id, connreq.peer_addr(), auth_id);
+        state.set_client_info(new_client_info);
+        let client_id = connreq.client_id().clone();
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:816]:
```suggestion
        
```

File: mqtt/mqtt-broker/src/broker.rs [hunk 16]
```diff
@@ -806,35 +806,36 @@ where
...
-            let client_id = connreq.client_id().clone();
-            let (state, _will, handle) = current_connected.into_parts();
-            let old_session = Session::new_disconnecting(state.client_info().clone(), None, handle);
-            let (new_session, session_present) = if let proto::ClientId::IdWithExistingSession(_) =
-                connreq.connect().client_id
-            {
+        let (mut state, _will, handle) = current_connected.into_parts();
+        let old_session = Session::new_disconnecting(state.client_info().clone(), None, handle);
+        let client_id = connreq.client_id().clone();
+        let new_client_info = ClientInfo::new(client_id, connreq.peer_addr(), auth_id);
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/broker.rs:814]: nit: 
```suggestion
        let new_client_info = ClientInfo::new(client_id.clone(), connreq.peer_addr(), auth_id);
```

File: mqtt/mqtt-broker/src/session/mod.rs [hunk 17]
```diff
@@ -27,17 +26,15 @@ pub enum Session {
 }
 
 impl Session {
-    pub fn new_transient(auth_id: AuthId, connreq: ConnReq, state: SessionState) -> Self {
-        let (client_id, peer_addr, connect, handle) = connreq.into_parts();
-        let client_info = ClientInfo::new(client_id, peer_addr, auth_id);
-        let connected = ConnectedSession::new(state, client_info, connect.will, handle);
+    pub fn new_transient(connreq: ConnReq, state: SessionState) -> Self {
```
Comment (dmolokanov) [inline mqtt/mqtt-broker/src/session/mod.rs:29]: nit: I wonder if we can just pass will message into both methods instead of ConnReq

Comment (dmolokanov) [reply mqtt/mqtt-broker/src/session/mod.rs:29]: if you do that you may not need to keep getters for `client_id` and `peer_addr` in the ConnReq

**Review by dmolokanov — APPROVED** [review]
Some nits, otherwise look good!

File: mqtt/mqtt-broker/src/broker.rs (hunk 16 above)
Comment (dylanbronson) [reply mqtt/mqtt-broker/src/broker.rs:814]: Clippy tells me not to

File: mqtt/mqtt-broker/src/session/mod.rs (hunk 17 above)
Comment (vadim-kovalyov) [reply mqtt/mqtt-broker/src/session/mod.rs:29]: I suggested to @dylanbronson to create another type `SessionReq` and convert ConnReq into SessionReq. That can help simplify code and types, but we agreed to try that later in a separate PR.

**Review by vadim-kovalyov — DISMISSED** [review]

File: mqtt/mqtt-broker/src/broker.rs (hunk 3 above)
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/broker.rs]: when we convert from proto::SubscribeTo to auth::Subscribe

File: mqtt/mqtt-broker/src/session/mod.rs (hunk 17 above)
Comment (dmolokanov) [reply mqtt/mqtt-broker/src/session/mod.rs:29]: Just curious what will be the purpose of `SessionReq`?

**Review by dmolokanov — DISMISSED** [review]
Now the code is much cleaner! Thanks!

**Review by vadim-kovalyov — APPROVED** [review]
